### PR TITLE
raft::read_barrier, unavailable leader handling

### PR DESCRIPTION
**Pull request by gusev-p**
If the leader is unavailable during read_barrier, closed_error occurs, which was not handled in any way and eventually reached the client. This patch adds retries in this case.

The patch also adds logging of server errors occuring while processing a request, as otherwise it is difficult to debug problems.

Fix: scylladb/scylla-dtest#2753

### CODE REVIEW

**Review by kostja — APPROVED** [review]

**Comment by gusev-p** [comment]
- added "continue;" when we got closed_error

**Comment by kbr-** [comment]
Is the CI failure related to the PR?

**Comment by gusev-p** [comment]
> Is the CI failure related to the PR?

Yes, I'm dealing with it right now. I've added error logging, and some test start to fail, because an error was found in the log. This error should not have happened according to the logic of the test.

**Comment by kbr-** [comment]
The commit message looks off.

File: raft/server.cc [hunk 1]
```diff
@@ -1142,7 +1143,14 @@ future<> server_impl::read_barrier(seastar::abort_source* as) {
             leader = _fsm->current_leader();
         } else {
             auto applied = _applied_idx;
-            auto res = co_await get_read_idx(leader, as);
+            read_barrier_reply res;
+            try {
+                res = co_await get_read_idx(leader, as);
+            } catch (const seastar::rpc::closed_error& e) {
+                logger.trace("[{}] read_barrier on {} resulted in seastar::rpc::closed_error; retrying", _id, leader);
+                leader = server_id{};
```
Comment (gleb-cloudius) [inline raft/server.cc:1151]: We may enter a busy loop here until leader changes. We need to sleep, but we raft code to be clock neutral.

**Comment by gusev-p** [comment]
splitted to separate, more focused PR https://github.com/scylladb/scylladb/pull/11263

issue https://github.com/scylladb/scylladb/issues/11262

**Pull request closed without merging**